### PR TITLE
[MLIR][TORCH] Add value tensor variant to aten::copy_ op

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit adds the op `ValsemVariantAtenCopyOp` that represents
`AtenCopy_Op` without the underscore. This is needed to make sure
that the `ReduceOpVariants` pass turns the in-place op into an op
that takes value tensors as inputs, otherwise the
`MaximizeValueSemantics` pass will not be able to add value
semantics correctly.

This commit also adds the lowering of `ValsemVariantAtenCopyOp`.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 1]
```diff
@@ -878,6 +878,88 @@ class ConvertAtenCatOp : public OpConversionPattern<AtenCatOp> {
...
 } // namespace
 
+// Broadcasts input tensor based on the broadcastToShape.
+template <typename OpTy>
+static LogicalResult
+broadcastToGivenShape(OpTy op, ConversionPatternRewriter &rewriter, Value input,
+                      SmallVector<Value> broadcastToShape, Value &result) {
+  RankedTensorType inputType = input.getType().cast<RankedTensorType>();
+  ArrayRef<int64_t> inputShape = inputType.getShape();
+  if (broadcastToShape.size() < inputShape.size())
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:888]: `if` needs curly braces

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 2]
```diff
@@ -878,6 +878,88 @@ class ConvertAtenCatOp : public OpConversionPattern<AtenCatOp> {
 };
 } // namespace
 
+// Broadcasts input tensor based on the broadcastToShape.
+template <typename OpTy>
+static LogicalResult
+broadcastToGivenShape(OpTy op, ConversionPatternRewriter &rewriter, Value input,
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:884]: I don't think this needs to be a template. I also noticed that there is nothing in the body of the function that requires information about the `op`. The `op` is only used to get a location and a context, however you can get a context from the rewriter itself. Here, I would use a similar signature to other helper functions

```cpp
broadcastToGivenShape(PatternRewriter &rewriter, Location loc, ...)
```

This makes it clear to the user that what this function does is `op`-independent.

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 3]
```diff
@@ -992,6 +1009,73 @@ class ConvertAtenContiguousOp : public OpConversionPattern<AtenContiguousOp> {
...
+    }
+
+    // The size of the src tensor can be different from the self but should be
+    // broadcastable. Therefore, broadcasting the src tensor to match the size
+    // of the self tensor.
+    SmallVector<Value> selfSizes = getTensorSizes(rewriter, loc, self);
+    for (unsigned i = 0; i < selfSizes.size(); i++)
+      selfSizes[i] = castIndexToInt(rewriter, loc, selfSizes[i]);
+    Value broadcastedSrc;
+    if (failed(broadcastToGivenShape<ValsemVariantAtenCopyOp>(
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:1047]: `if` needs curly braces

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Thanks for refactoring the broadcast code!

File: lib/Conversion/TorchToLinalg/DataMovement.cpp (hunk 2 above)
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:884]: Never mind, you do need the `op` as an argument to create the `notifyMatchFailure`. So you should keep the signature unchanged. However, I would remove the template code and pass the `op` as `Operation *op`.

**Review by ramiro050 — APPROVED** [review]
LGTM! Just one small comment

**Review by pashu123 — APPROVED** [review]
LGTM!

**Pull request merged**